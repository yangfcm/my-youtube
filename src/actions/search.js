import { defAxios as axios } from "../settings";
import { SEARCH, CATCH_ERROR } from "./types";
import { DEFAULT_ERROR_MSG } from "./default-error-msg";

/**
 * Search resources (including videos, channels, play lists)
 */
export const searchVideos = (filter, pageToken) => {
  return async (dispatch) => {
    try {
      const response = await axios.get("/search", {
        params: {
          ...axios.defaults.params,
          ...filter,
          part: "snippet",
          maxResults: 15,
          pageToken,
        },
      });
      dispatch({
        type: SEARCH,
        payload: response.data,
      });
    } catch (e) {
      dispatch({
        type: CATCH_ERROR,
        payload: {
          ...e.response.data.error,
          displayMessage: DEFAULT_ERROR_MSG.FAILED_TO_SERCH,
        },
      });
    }
  };
};
