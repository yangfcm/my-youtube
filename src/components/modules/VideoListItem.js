import React from "react";
import { Link } from "react-router-dom";
import moment from "moment";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

const VideoListItem = ({ video, playlistId }) => {
  // console.log(video);
  let linkUrl = video.contentDetails
    ? `/video/${video.contentDetails.videoId}`
    : `/video/${video.id.videoId}`;
  if (playlistId) linkUrl += `?playlistId=${playlistId}`;
  return (
    <div className="d-flex flex-md-row my-2">
      <div className="d-flex align-items-center">
        <Link to={linkUrl}>
          <img
            src={
              video.snippet.thumbnails && video.snippet.thumbnails.medium.url
            }
            alt={video.title}
            style={{ maxWidth: "140px" }}
          />
        </Link>
      </div>
      <div className="d-flex flex-column justify-content-center ml-3">
        <Link to={linkUrl} title={video.snippet.title}>
          <div
            className="text-primary font-weight-bold"
            style={{
              lineHeight: "23px",
              maxHeight: "46px",
              overflow: "hidden",
              textOverflow: "ellipsis",
            }}
          >
            {video.snippet.title}
          </div>
        </Link>
        <Link to={`/channel/${video.snippet.channelId}`}>
          {" "}
          {video.snippet.channelTitle}
        </Link>
        <p className="font-weight-light">
          <FontAwesomeIcon icon="user-clock" />{" "}
          {moment(video.snippet.publishedAt).format("D MMM YYYY k:mm")}
        </p>
      </div>
    </div>
  );
};

export default VideoListItem;
