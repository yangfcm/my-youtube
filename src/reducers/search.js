import { SEARCH } from "../actions/types";

export const searchReducer = (state = {}, action) => {
  switch (action.type) {
    case SEARCH:
      return {
        ...state,
        searchResults: action.payload,
      };
    default:
      return state;
  }
};
