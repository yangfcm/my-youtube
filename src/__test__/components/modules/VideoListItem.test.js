import React from "react";
import { shallow } from "enzyme";
import { Link } from "react-router-dom";
import moment from "moment";
import VideoListItem from "components/modules/VideoListItem";
import { channelVideosResponse } from "__test__/fixtures/search";

describe("Test VideoListItem component", () => {
  let wrapper;
  const video = channelVideosResponse.items[0];
  const playlistId = "playlist-id";
  beforeEach(() => {
    wrapper = shallow(<VideoListItem video={video} playlistId={playlistId} />);
  });

  afterEach(() => {
    wrapper.unmount();
  });

  it("should render component", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render correct video image", () => {
    expect(wrapper.find("img").prop("src")).toBe(
      video.snippet.thumbnails.medium.url
    );
  });

  it("should render correct links to that navigate to video and channel", () => {
    const links = wrapper.find(Link);
    expect(links.at(0).prop("to")).toBe(
      `/video/${video.id.videoId}?playlistId=${playlistId}`
    );
    expect(links.at(1).text()).toContain(video.snippet.title);
    expect(links.at(1).prop("to")).toBe(
      `/video/${video.id.videoId}?playlistId=${playlistId}`
    );
    expect(links.at(2).text()).toContain(video.snippet.channelTitle);
    expect(links.at(2).prop("to")).toBe(`/channel/${video.snippet.channelId}`);
  });

  it("should render correct links if playlist id is not given", () => {
    const wrapperWithoutPlaylist = shallow(<VideoListItem video={video} />);
    const links = wrapperWithoutPlaylist.find(Link);
    expect(links.at(0).prop("to")).toBe(`/video/${video.id.videoId}`);
    expect(links.at(1).prop("to")).toBe(`/video/${video.id.videoId}`);
  });

  it("should render correct publish date time", () => {
    expect(wrapper.find("p.font-weight-light").text()).toContain(
      moment(video.snippet.publishedAt).format("D MMM YYYY k:mm")
    );
  });
});
