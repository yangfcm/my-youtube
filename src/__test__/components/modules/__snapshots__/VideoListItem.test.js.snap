// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test VideoListItem component should render component 1`] = `
<div
  className="d-flex flex-md-row my-2"
>
  <div
    className="d-flex align-items-center"
  >
    <Link
      to="/video/McOW1uIE4GY?playlistId=playlist-id"
    >
      <img
        src="https://i.ytimg.com/vi/McOW1uIE4GY/mqdefault.jpg"
        style={
          Object {
            "maxWidth": "140px",
          }
        }
      />
    </Link>
  </div>
  <div
    className="d-flex flex-column justify-content-center ml-3"
  >
    <Link
      title="Coronavirus: UN warns of new diseases - News Review"
      to="/video/McOW1uIE4GY?playlistId=playlist-id"
    >
      <div
        className="text-primary font-weight-bold"
        style={
          Object {
            "lineHeight": "23px",
            "maxHeight": "46px",
            "overflow": "hidden",
            "textOverflow": "ellipsis",
          }
        }
      >
        Coronavirus: UN warns of new diseases - News Review
      </div>
    </Link>
    <Link
      to="/channel/UCHaHD477h-FeBbVh9Sh7syA"
    >
       
      BBC Learning English
    </Link>
    <p
      className="font-weight-light"
    >
      <FontAwesomeIcon
        border={false}
        className=""
        fixedWidth={false}
        flip={null}
        icon="user-clock"
        inverse={false}
        listItem={false}
        mask={null}
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        swapOpacity={false}
        symbol={false}
        title=""
        transform={null}
      />
       
      4 Aug 2020 21:30
    </p>
  </div>
</div>
`;
